addons/pos_online_payment/controllers/payment_portal.py

```diff
@@ -0,0 +1,299 @@
+
+    @staticmethod
+    def _get_exit_route_arg(exit_route):
+        return f'&exit_route={urllib.parse.quote(exit_route)}' if exit_route else ''
+
+    @staticmethod
+    def _get_pay_route(pos_order_id, access_token, exit_route=None):
+        exit_route_arg = PaymentPortal._get_exit_route_arg(exit_route)
+        return f'/pos/pay/{pos_order_id}?access_token={access_token}{exit_route_arg}'
+
+    @staticmethod
+    def _get_landing_route(pos_order_id, access_token, exit_route_arg=None, tx_id=None):
+        tx_id_arg = f'&tx_id={tx_id}' if tx_id else ''
+        return f'/pos/pay/confirmation/{pos_order_id}?access_token={access_token}{exit_route_arg}{tx_id_arg}'
+
+    @http.route('/pos/pay/<int:pos_order_id>', type='http', methods=['GET'], auth='public', website=True, sitemap=False)
+    def pos_order_pay(self, pos_order_id, access_token=None, exit_route=None):
+        """ Behaves like payment.PaymentPortal.payment_pay but for POS online payment.
+
+        :param str pos_order_id: The POS order to pay, as a `pos.order` id
+        :param str access_token: The access token used to verify the user
+        :param str exit_route: The URL to open to leave the POS online payment flow
+
+        :return: The rendered payment form
+        :rtype: str
+        :raise: AccessError if the provided order or access token is invalid
+        :raise: ValidationError if data on the server prevents the payment
+        """
+        pos_order_sudo = self._check_order_access(pos_order_id, access_token)
+        self._ensure_session_open(pos_order_sudo)
+
+        user_sudo = request.env.user
+        logged_in = not user_sudo._is_public()
+        partner_sudo = self._get_partner_sudo(user_sudo)
+        if not partner_sudo:
+            return self._redirect_login(request)
+
+        kwargs = {
+            'pos_order_id': pos_order_sudo.id,
+        }
+        exit_route_arg = self._get_exit_route_arg(exit_route)
+        rendering_context = {
+            **kwargs,
+            'exit_route': exit_route,
+            'reference_prefix': request.env['payment.transaction'].sudo()._compute_reference_prefix(provider_code=None, separator='-', **kwargs),
+            'partner_id': partner_sudo.id,
+            'access_token': access_token,
+            'transaction_route': f'/pos/pay/transaction/{pos_order_sudo.id}?access_token={access_token}{exit_route_arg}',
+            'landing_route': self._get_landing_route(pos_order_sudo.id, access_token, exit_route_arg=exit_route_arg),
+            **self._get_custom_rendering_context_values(**kwargs),
+        }
+
+        currency_id = pos_order_sudo.currency_id
+
+        if not self._is_valid_currency(currency_id):
+            rendering_context['currency'] = False
+            return self._render_pay(request, rendering_context)
+        rendering_context['currency'] = currency_id
+
+        amount_to_pay = self._get_amount_to_pay(pos_order_sudo)
+        if not self._is_valid_amount(amount_to_pay, currency_id):
```

Comment:
I'm not sure to understand, but you would like that _get_amout_to_pay returns the amount if it is valid, or something like False otherwise ?
So I would need to check if the result is a boolean, and check if it is False to determine if the amount is invalid and then display it on the payment page.
I find my approach easier to understand, it clearly highlights that I get the amount that should be paid, and then I check if it is valid for the payment portal page.

Reply:
I totally agree with you, except one thing: this is not a method intended to be used outside of the payment portal. This is an internal method only intended for the payment portal. It is a method to highlight that is is used several times, and guarantee that it will always produce the same result.
If you want the amount to pay of an order, you can call the method in pos.order. Nobody should and would use the payment portal to get the amount to pay of an order.